src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java

```diff
@@ -107,29 +109,47 @@ public static boolean isCigarValid(Cigar cigar) {
         return false;
     }
 
-    public static final int countRefBasesBasedOnCigar(final SAMRecord read, final int cigarStartIndex, final int cigarEndIndex){
+    /**
+     * Compute the number of reference bases between the start (inclusive) and end (exclusive) cigar elements.
+     */
+    @SuppressWarnings("fallthru")
+    public static int countRefBasesBasedOnCigar(final SAMRecord read, final int cigarStartIndex, final int cigarEndIndex){
+        if (read == null){
+            throw new IllegalArgumentException("null read");
+        }
+        final List<CigarElement> elems = read.getCigar().getCigarElements();
+        if (cigarStartIndex < 0 || cigarEndIndex > elems.size() || cigarStartIndex > cigarEndIndex){
+            throw new IllegalArgumentException("invalid index:" + 0 + " -" + elems.size());
+        }
         int result = 0;
-        for(int i = cigarStartIndex; i<cigarEndIndex;i++){
-            final CigarElement cigarElement = read.getCigar().getCigarElement(i);
+        for(int i = cigarStartIndex; i < cigarEndIndex; i++){
+            final CigarElement cigarElement = elems.get(i);
             switch (cigarElement.getOperator()) {
                 case M:
-                case S:
                 case D:
                 case N:
-                case H:
+                case EQ:
+                case X:
+                case S:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
+                case P:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
+                case H:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
                     result += cigarElement.getLength();
                     break;
-                case I:
+                default://I
                     break;
-                default:
-                    throw new GATKException("Unsupported cigar operator: " + cigarElement.getOperator());
             }
         }
         return result;
     }
 
-    public static Cigar unclipCigar(Cigar cigar) {
-        ArrayList<CigarElement> elements = new ArrayList<>(cigar.numCigarElements());
+    /**
+     * Removes all clipping operators from the cigar.
+     */
+    public static Cigar unclipCigar(final Cigar cigar) {
```

Comment:
it is not the right name for the function, I thought unclipping is to take the clipped bases and mark them as M instead. I would call it trimReadToUnclippedBases.